### ISSUE TITLE
[tune] pytorch-lightning not using gpu

### ISSUE THREAD

**Issue by Data-drone** [ray-project/ray#13311]
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*

Ray 1.1.0
Torch 1.7.0a0
torchvision 0.8.0
pytorch-lightning 1.1.2
ubuntu 20.04

### Reproduction (REQUIRED)

See: https://gist.github.com/Data-drone/b48a0cc1026ec40a6901480aaec32f6b

Running this script in an environment as per the above, the pytorch training doesn't seem to be leveraging GPUs in training the neural network.

**Comment by amogkam** [comment]
Hi @Data-drone do you see something like this being printed out by PTL in the console output?
```
(pid=281) /usr/local/lib/python3.6/dist-packages/pytorch_lightning/utilities/distributed.py:49: UserWarning: You have set progress_bar_refresh_rate < 20 on Google Colab. This may crash. Consider using progress_bar_refresh_rate >= 20 in Trainer.
(pid=281)   warnings.warn(*args, **kwargs)
(pid=281) GPU available: True, used: True
(pid=281) TPU available: None, using: 0 TPU cores
(pid=281) LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
```

**Comment by Data-drone** [comment]
I see:

```
2021-01-11 02:38:17,927 WARNING function_runner.py:539 -- Function checkpointing is disabled. This may result in unexpected behavior when using checkpointing features or certain schedulers. To enable, set the train function arguments to be `func(config, checkpoint_dir=None)`.
2021-01-11 02:38:18,190 ERROR syncer.py:72 -- Log sync requires rsync to be installed.
== Status ==
Memory usage on this node: 15.9/62.8 GiB
Using AsyncHyperBand: num_stopped=0
Bracket: Iter 8.000: None | Iter 4.000: None | Iter 2.000: None | Iter 1.000: None
Resources requested: 4/16 CPUs, 1/2 GPUs, 0.0/36.13 GiB heap, 0.0/12.45 GiB objects (0/1.0 accelerator_type:RTX)
Result logdir: /home/jovyan/ray_results/tune_mnist_asha
Number of trials: 1/10 (1 RUNNING)
+--------------------+----------+-------+----------------+----------------+------------+--------------+
| Trial name         | status   | loc   |   layer_1_size |   layer_2_size |         lr |   batch_size |
|--------------------+----------+-------+----------------+----------------+------------+--------------|
| _inner_0fd70_00000 | RUNNING  |       |             32 |            128 | 0.00176656 |           32 |
+--------------------+----------+-------+----------------+----------------+------------+--------------+


(pid=1109) GPU available: True, used: True
(pid=1109) TPU available: None, using: 0 TPU cores
(pid=1109) LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
(pid=1110) GPU available: True, used: True
(pid=1110) TPU available: None, using: 0 TPU cores
(pid=1110) LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [1]
(pid=1110) 
(pid=1110)   | Name    | Type   | Params
(pid=1110) -----------------------------------
(pid=1110) 0 | layer_1 | Linear | 25.1 K
(pid=1110) 1 | layer_2 | Linear | 4.2 K 
(pid=1110) 2 | layer_3 | Linear | 1.3 K 
(pid=1110) -----------------------------------
(pid=1110) 30.6 K    Trainable params
(pid=1110) 0         Non-trainable params
(pid=1110) 30.6 K    Total params
(pid=1110) /opt/conda/envs/computer_vision/lib/python3.8/site-packages/pytorch_lightning/utilities/distributed.py:49: UserWarning: The dataloader, val dataloader 0, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 16 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
(pid=1110)   warnings.warn(*args, **kwargs)
```

**Comment by amogkam** [comment]
Hmm based off the `(pid=1109) GPU available: True, used: True` line, Pytorch Lightning is showing that GPU is being used.

When you no longer use Ray and just use Pytorch Lightning instead, do you see GPU being utilized?
Also how are you measuring this utilization? Could you also share some output from this as well?

**Comment by Data-drone** [comment]
I think maybe it is just being underutilised. I will have to build up a bigger test that pushes gpu more. Maybe switch for fashion mnist or something and switch network for a resnet50

**Comment by Data-drone** [comment]
okay with this gist: https://gist.github.com/Data-drone/4176c6c1fcf2b573ae8009b7a9ce53f5
if you run `train_mnist(single_config)` I am getting about 30% gpu utilisation but with `tune_mnist_asha(num_samples=10, num_epochs=50, gpus_per_trial=1, cpus_per_trial=4)` the two gpus I have in the system don't go about 10% or so. One is usually around 3-5%

**Comment by Data-drone** [comment]
okay with this one: https://gist.github.com/Data-drone/d5171f3dbceade3917ef34bb5860ee0b
if you run the pure pytorch-lightning script I get 60%+ utilisation on GPU whereas with the asha code I am seeing 10% to 20% less utilisation than before.

**Comment by amogkam** [comment]
@Data-drone thanks for the update- interesting to see those numbers.

To clarify, for the `train_mnist` case with a single config, are you calling it directly and removing the `TuneReportCallback`, or are you still invoking it via `tune.run`, but just not using the ASHA scheduler?

cc @krfricke @richardliaw do you guys have an idea on what could be causing this difference in utilization?

**Comment by Data-drone** [comment]
the single `train_mnist` case doesn't have the `TuneReportCallBack` in it. so `train_mnist` is "stock lightning" so to speak

So my tests have been to run the `train_mnist` function to see how much GPU usage I am getting then to run the `tune_mnist_asha` function to run it with ray. I may not understand the `tune_mnist_asha` function correctly but by setting `gpus_per_trial=1` and `num_epochs=50` that should mean that each trial has access to a gpu and it will run for enough epochs to see some steady gpu utilisation prior to asha terminating underperforming trials? Or am I mistaken?

**Comment by amogkam** [comment]
Hey @Data-drone, could you also try running training with Tune (still use the `TuneReportCallback` and run it with `tune.run`), but don't use any scheduler. That'll help with identifying if the ASHA scheduler is what's causing the issue or if its some general Tune overhead.

**Comment by Data-drone** [comment]
Okay I just tried https://gist.github.com/Data-drone/d5171f3dbceade3917ef34bb5860ee0b again with the `scheduler` commented out from the `tune.run`. GPU is hovering about the 40%-50% mark so similar to when `asha` was on. Could it be that the `TuneReportCallback` or the  `CLIReporter` is taking up cycles and leaving GPU waiting?

**Comment by amogkam** [comment]
Hey @Data-drone can you also try with `num_samples=1` instead of 10/

**Comment by amogkam** [comment]
Also can you remove `batch_size` as a hyperparameter. @richardliaw pointed out that this could be the cause for the discrepancy.

**Comment by Data-drone** [comment]
Hmm I took out batch size and dropped num_samples and still seem to be seeing lower than expected GPU usage. I did try to run a profile on it with scalene: https://github.com/emeryberger/scalene but cause of the way executes it doesn't give much useful info.

What is the expected perf drop of using Ray versus standalone? Are there any dev estimates. I also haven't run a cluster setup so maybe that changes things?

**Comment by richardliaw** [comment]
Oh one possibility is that ray automatically sets `OMP_NUM_THREADS` on each worker. Could you try adding a line like:

```
os.environ["OMP_NUM_THREADS"] = 4
```

**Comment by Data-drone** [comment]
What does 0MP thread do?

**Comment by richardliaw** [comment]
https://www.openmp.org/spec-html/5.0/openmpse50.html

**Comment by dvirginz** [comment]
I'm following along, as I have the exact same problem.
Running vanilla pl I have 80% utilization, running from ray I have 3%.

I believe this has something to do we newer versions of ray/py, as it doesn't happen to me only after building my env from scrach.

**Comment by richardliaw** [comment]
@dvirginz, could you post a reproduction script? It'd be great if you could let me know what type of machines you are on though. Also, are you using conda?

**Comment by richardliaw** [comment]
@Data-drone I took a look at the script https://gist.github.com/Data-drone/d5171f3dbceade3917ef34bb5860ee0b and commented out the batch size parameter -- 

With tune + without Tune seems the same for me (running on AWS machine) in terms of GPU utilization.

@dvirginz if you have a reproduction script I'd love to take a closer look; unfortunately I can't find any issues with whatever is given in the thread (after commenting out batch size), so I'll close this issue for now.
